ggsql announcement (#170)
posit-dev/open-source-website: content/blog/ggsql_alpha_release/index.md

---
title: 'ggsql: A grammar of graphics for SQL'
date: 2026-04-17T00:00:00.000Z
people:
  - Thomas Lin Pedersen
  - Teun Van den Brand
  - George Stagg
  - Hadley Wickham
description: >
  Introducing ggsql, a grammar of graphics for SQL that lets you describe
  visualizations directly inside SQL queries.
image: ggsql.png
image-alt: >-
  ggsql logo. A pencil with the body consisting of the standard database silo
  icon.
topics:
  - Visualization
software:
  - ggsql
languages:
  - SQL
nohero: false
hidesubscription: false
---

<script src="https://cdn.jsdelivr.net/npm/requirejs@2.3.6/require.min.js" integrity="sha384-c9c+LnTbwQ3aujuU7ULEPVvgLs+Fn6fJUvIGTsuu1ZcCf11fiEubah0ttpca4ntM sha384-6V1/AdqZRWk1KAlWbKBlGhN7VG4iE/yAZcO6NZPMF8od0vukrvr0tg4qY6NSrItx" crossorigin="anonymous"></script>
<script src="https://cdn.jsdelivr.net/npm/jquery@3.5.1/dist/jquery.min.js" integrity="sha384-ZvpUoO/+PpLXR1lu4jmpXWu80pZlYUAfxl5NsBMWOEPSjUn/6Z/hRTt8+pR6L4N2" crossorigin="anonymous" data-relocate-top="true"></script>
<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>

> Today, we are super excited to formally announce the alpha-release of [ggsql](https://ggsql.org). As the name suggests, ggsql is an implementation of the grammar of graphics based on the SQL syntax, bringing rich, structured visualization support to SQL for the first time.

Reviewer: ```suggestion
Today, we are super excited to announce the alpha-release of [ggsql](https://ggsql.org). As the name suggests, ggsql is an implementation of the grammar of graphics based on SQL syntax, bringing rich, structured visualization support to SQL.
```

> In this post we will go over some of the motivations that lead us to develop this tool, as well as give you ample examples of its use; so you can hopefully get as excited about it as we are (if you can't wait to get excited, [jump straight to the examples](#meet-ggsql)).

Reviewer: IMO you should include one example here

Reviewer: Or even flip the order; show the results then discuss the motivation.

## Why ggsql?

> Writing a new visualization library from scratch is always a big task and you may wonder why we have decided on this endavour. Some of the reasons are:

Reviewer: ```suggestion
Writing a new visualization library from scratch is a big task and you might wonder why we're doing it again. Some of the reasons are:
```

- We want to engage with and help data analysts and data scientists that predominantly work in SQL

> - SQL and the grammar of graphics just fits together extremely well

Reviewer: ```suggestion
- SQL and the grammar of graphics fit together extremely well
```

> - We want to create an extremely powerful, code-based, visualization tool that doesn't require a runtime

Reviewer: ```suggestion
- We want to create an extremely powerful, code-based, visualization tool that doesn't require an entire programming language (like R or python)
```

> - We have learned so much after 18 years of [ggplot2](https://ggplot2.tidyverse.org/) development and maintenance that it feels right to apply it to a blank canvas

Reviewer: ```suggestion
- We have learned so much from 18 years of [ggplot2](https://ggplot2.tidyverse.org/) development that we're excited to apply to a blank canvas
```

Let's discuss each in turn.

### Hello, SQL user!

<img src="greet.png" alt="Humorous SQL query greeting SQL users that aren't R and Python users as well: SELECT 'Hello' AS greetings FROM SQL_users WHERE other_language NOT IN ('R', 'Python')" />

> While first R and then Python captured all the attention of the data science revolution, SQL chucked along as the reliant and powerful workhorse beneath it all. We recognize that there are large portions of people that work with data that do so only or predominantly in SQL. The choice they have for visualizing their data are often suboptimal in our view:

Reviewer: ```suggestion
While first R and then Python captured all the attention of the data science revolution, SQL chugged along as the reliable and powerful workhorse beneath it all. There are many people who work with data that do so only or predominantly in SQL. The choice they have for visualizing their data are often suboptimal in our view:
```

<img src="epic_handshake.png" alt="Epic handshake meme showing SQL and the grammar of graphics agreeing on declarative, composable operations" />

From the above, slightly hyperbolic, overview it is clear that both SQL and the grammar of graphics have a lot of commonality in their approach to their respective domains. Together they can offer a very powerful and natural solution to the full pipeline from raw data to final visualization.

### No runtime, no problem

Why does it matter that [ggplot2](https://ggplot2.tidyverse.org/) and [plotnine](https://plotnine.org) requires R and Python installed respectively? There are clear benefits to a single, focused executable to handle data visualization:

- Embedding a small executable in other tools is much easier than bundling R/Python (or requiring them to be installed)

> - Clear scope and need makes it easier to sandbox and prevent malicious code execution (either deliberately or in error)

Reviewer: ```suggestion
- A smaller scope makes it easier to sandbox and prevent malicious code execution (either deliberately or in error)
```

Both of the above points make ggsql a much more compelling option for integrating into tools such as AI agents assisting you in data analysis, or code based reporting tools that may execute code in different environments.

> You may think we have had to swallow some bitter pills by moving away from an interpreted language, but it has also given us a lot. Most importantly, the rigid structure means that we can execute the whole data pipeline as a single SQL query per layer on the backend. This means that if you want to create a bar plot of 10 billion transactions you only ever fetch the count values for each bar from your data warehouse, not the 10 billion rows of data. The same is true for more complicated layer types such as boxplots and density plots. This is in stark contrast to the situation in most visualization tools that first materializes the data (potentially choking on the size), then perform the necessary computations on it, before proceeding to plot it.

Reviewer: ```suggestion
You may think we have had to swallow some bitter pills by moving away from an interpreted language, but it has also given us a lot. Most importantly, the rigid structure means that we can execute the whole data pipeline as a single SQL query per layer on the backend. This means that if you want to create a bar plot of 10 billion transactions you only ever fetch the count values for each bar from your data warehouse, not the 10 billion rows of data. The same is true for more complicated layer types such as boxplots and density plots. This is in stark contrast to most visualization which must first materialize the complete data, then perform the necessary computations on it, then plot it.
```

### We are now wise beyond our years

<img src="age.gif" style="display: block; margin: auto;" alt="Titanic meme showing aged Rose saying 'It's been 84 years'" />

> 18 years of ggplot2 development and maintenance also means 18 years of thinking about data visualization syntax, use, and design. I am not a boastful person but I do believe that gives us some expert knowledge on the subject matter. However, not all of this knowledge can be poured back into ggplot2. There are decisions and expectations established many years ago that we have to honor, or at least only challenge very gradually (which we do on occasions).

Reviewer: "I" doesn't work with multi-authors

ggsql is a blank slate. Not only in the sense that we are building it from the ground up, but also in that it is built for an environment with no established expectations for a visualization tool. I cannot stress how liberating and invigorating this has felt, and I am positive that this shines through in how ggsql feels to the user.

## Meet ggsql

That was a lot of words about the why. Hopefully you come away from it all with a sense of this whole thing being justified, ready to learn more about it. If so, read on!

### A (simple) example

Let's start with some code even though we haven't introduced the syntax yet. I promise we will go through it below. The example is an adaptation of a visualization created by [Jack Davison](https://jack-davison.github.io) for TidyTuesday.

``` ggsql
WITH astronauts AS (
  SELECT * FROM 'astronauts.parquet'
  QUALIFY ROW_NUMBER() OVER (
    PARTITION BY name
    ORDER BY mission_number DESC

>   ) = 1

Reviewer: Could we simplify this? It's a bit distracting to hit some complex SQL syntax when the focus should be on the visualisation.

Author: If the example is moved down don't you think it will be ok. The point was really to show a full pipeline including some meaningful SQL preparation

'vega': 'https://cdn.jsdelivr.net/npm/vega@6/build/vega.min',
'vega-lite': 'https://cdn.jsdelivr.net/npm/vega-lite@6.4.1/build/vega-lite.min',
'vega-embed': 'https://cdn.jsdelivr.net/npm/vega-embed@7/build/vega-embed.min'
}
});
function docReady(fn) {
if (document.readyState === 'complete') fn();
else window.addEventListener("load", function() { fn(); });
}
docReady(function() {
window.requirejs(["dom-ready", "vega", "vega-embed"], function(domReady, vega, vegaEmbed) {
domReady(function () {
vegaEmbed('#' + visId, spec, options).then(onRendered).catch(console.error);
});
});
});
} else {
function loadScript(src) {
return new Promise(function(resolve, reject) {
var script = document.createElement('script');
script.src = src;
script.onload = resolve;
script.onerror = reject;
document.head.appendChild(script);
});
}
Promise.all([
loadScript('https://cdn.jsdelivr.net/npm/vega@6'),
loadScript('https://cdn.jsdelivr.net/npm/vega-lite@6.4.1'),
loadScript('https://cdn.jsdelivr.net/npm/vega-embed@7')
])
.then(function() { return vegaEmbed('#' + visId, spec, options); })
.then(onRendered)
.catch(function(err) {
console.error('Failed to load Vega libraries:', err);
});
}
})();
</script>

> That was a lot of code, but on the flip-side it allows us to cover a lot of the most important aspects of the syntax with one example.

Reviewer: ```suggestion
That was a lot of code, but it allows us to cover a lot of the most important aspects of the syntax with one example.
```

At the topmost level there are two parts to an SQL query: The SQL query, and the visualization query. The SQL query is anything from the beginning to the `VISUALIZE` clause. It is your standard SQL, and it accepts anything your backend accepts (in this blog post we use a DuckDB backend). The result of the query is funnelled directly into the visualization rather than being returned as a table like you'd normally expect.

Since the point of this post is not to teach you SQL we won't spend much more time discussing the SQL query part. The main take away is that everything before the `VISUALIZE` clause is pure SQL, any resulting table is automatically used by your visualization, and any table or CTE created there is available for referencing in the visualization query.

It should also be noted that the SQL query part is optional. If your data is already in the right shape for plotting you can skip it and instead name the source directly in the `VISUALIZE` clause:

``` ggsql

> VISUALIZE year_of_selection AS x, year_of_mission AS y FROM 'astronauts.parquet'

Reviewer: ```suggestion
FROM 'astronauts.parquet'
VISUALIZE year_of_selection AS x, year_of_mission AS y 
```

?

Author: that won't work - everything above VISUALIZE is ignored

Reviewer 2: We probably should add support for the equivalent of duckdb's,

```
D FROM 'penguins.csv' LIMIT 2;
┌─────────┬───────────┬───┬───────────────────┬─────────────┬─────────┐
│ species │  island   │ … │ flipper_length_mm │ body_mass_g │   sex   │
│ varchar │  varchar  │   │       int64       │    int64    │ varchar │
├─────────┼───────────┼───┼───────────────────┼─────────────┼─────────┤
│ Adelie  │ Torgersen │ … │               181 │        3750 │ MALE    │
│ Adelie  │ Torgersen │ … │               186 │        3800 │ FEMALE  │
├─────────┴───────────┴───┴───────────────────┴─────────────┴─────────┤
│ 2 rows                                          7 columns (5 shown) │
└─────────────────────────────────────────────────────────────────────┘
```

if we can. I have had someone else ask for the same thing, with `FROM` at the front.

Reviewer 2: Maybe not for initial release, if the required grammar changes end up being a pain.
